### ISSUE TITLE
Regression: `@psalm-assert !empty` with `bool` does not reduce it to `true`

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#5049]
While writing type refinements for `webmozart/assert`, I found a regression between `vimeo/psalm@61f5a06` (61f5a06) and `vimeo/psalm:4.4.1` (9fd7a7).

https://psalm.dev/r/6d6857dcf8

Specifically:

```php
/** @psalm-assert !empty $input */
function assertNotEmpty(mixed $input): void { /* ... */ }

/** @psalm-return true */
function gimmeTrue(bool $input): bool {
    assertNotEmpty($input);
    
    return $input;
}
```

Produces:

```
Psalm output (using commit 2f58c6a): 

ERROR: InvalidReturnStatement - 10:12 - The inferred type 'bool' does not match the declared return type 'true' for gimmeTrue

ERROR: InvalidReturnType - 6:19 - The declared return type 'true' for gimmeTrue is incorrect, got 'bool'
 
Psalm detected 1 fixable issue(s)
```

When given a `bool` `$input`, I would expect this to infer `true` as result of the assertion: this used to work in `3.x` (61f5a06).

The opposite works:

```php
/** @psalm-assert empty $input */
function assertEmpty(mixed $input): void { /* ... */ }

/** @psalm-return false */
function gimmeTrue(bool $input): bool {
    assertEmpty($input);
    
    return $input;
}
```

https://psalm.dev/r/ff0756f05d

**Comment by Ocramius** [comment]
Fixed in #5050 (assuming that patch is correct)

**Comment by Ocramius** [comment]
Fixed in #5050 (assuming that patch is correct)
